File: native-kt-runtime-jvm/src/main/java/com/huskerdev/nativekt/jvmci/JVMCIUtils.java
package com.huskerdev.nativekt.jvmci;

import com.huskerdev.nativekt.NativeKtUtils;
import jdk.vm.ci.code.site.*;
import jdk.vm.ci.hotspot.*;
import jdk.vm.ci.meta.*;
import jdk.vm.ci.runtime.JVMCI;
import jdk.vm.ci.runtime.JVMCIBackend;
import java.lang.reflect.Method;

import static com.huskerdev.nativekt.NativeKtUtils.addExports;

public class JVMCIUtils {

    static {
        if(NativeKtUtils.isAutoExportEnabled()) {
            addExports(JVMCIUtils.class.getModule(), "jdk.internal.vm.ci", new String[] {
                    "jdk.vm.ci.code",
                    "jdk.vm.ci.runtime",
                    "jdk.vm.ci.meta",
                    "jdk.vm.ci.hotspot",
                    "jdk.vm.ci.code.site"
            });
            addExports("jdk.graal.compiler", "java.base", new String[] {
                    "jdk.internal.misc"
            });
        }
    }

    public static void linkNativeCall(Method method, long address){
        CallingConvention convention = CallingConvention.current();

        JVMCIBackend jvmci = JVMCI.getRuntime().getHostJVMCIBackend();
        HotSpotResolvedJavaMethod resolvedMethod = (HotSpotResolvedJavaMethod) jvmci.getMetaAccess().lookupJavaMethod(method);

        jvmci.getCodeCache().setDefaultCode(resolvedMethod, convention.createNMethod(
                method.getName(),
                convention.createNativeCall(method, address),
                resolvedMethod
        ));
    }
}

File: modules/runtime/src/jvmMain/kotlin/com/huskerdev/nativekt/jvm/jvmci/JVMCIUtils.kt
@file:Suppress("JAVA_MODULE_DOES_NOT_EXPORT_PACKAGE")

package com.huskerdev.nativekt.jvm.jvmci

import com.huskerdev.nativekt.jvm.NativeKtUtils
import jdk.vm.ci.hotspot.HotSpotResolvedJavaMethod
import jdk.vm.ci.runtime.JVMCI
import java.lang.reflect.Method

object JVMCIUtils {
    init {
        if(NativeKtUtils.isAutoExportEnabled()) {
            NativeKtUtils.addExports(
                JVMCIUtils::class.java.module,
                "jdk.internal.vm.ci",
                arrayOf(
                    "jdk.vm.ci.code",
                    "jdk.vm.ci.runtime",
                    "jdk.vm.ci.meta",
                    "jdk.vm.ci.hotspot",
                    "jdk.vm.ci.code.site"
                )
            )
            NativeKtUtils.addExports(
                "jdk.graal.compiler",
                "java.base",
                arrayOf("jdk.internal.misc")
            )
        }
    }

    val emptyCharArray = charArrayOf()
    val emptyBooleanArray = booleanArrayOf()
    val emptyByteArray = byteArrayOf()
    val emptyShortArray = shortArrayOf()
    val emptyIntArray = intArrayOf()
    val emptyLongArray = longArrayOf()
    val emptyFloatArray = floatArrayOf()
    val emptyDoubleArray = doubleArrayOf()

    fun linkNativeCall(method: Method, address: Long){
        val convention = CallingConvention.current

        val jvmci = JVMCI.getRuntime().hostJVMCIBackend
        val resolvedMethod = jvmci.metaAccess.lookupJavaMethod(method) as HotSpotResolvedJavaMethod

        jvmci.codeCache.setDefaultCode(
            resolvedMethod,
            convention.createNMethod(
                method.name,
                convention.createNativeCall(method, address),
                resolvedMethod
            )
        )
    }
}
